tokio/src/sync/set_once.rs

```diff
@@ -0,0 +1,371 @@
+        } else {
+            SetOnce::new()
+        }
+    }
+
+    /// Creates a new `SetOnce` that contains the provided value.
+    ///
+    /// # Example
+    ///
+    /// When using the `tracing` [unstable feature], a `SetOnce` created with
+    /// `const_new_with` will not be instrumented. As such, it will not be
+    /// visible in [`tokio-console`]. Instead, [`SetOnce::new_with`] should be
+    /// used to create an instrumented object if that is needed.
+    ///
+    /// ```
+    /// use tokio::sync::SetOnce;
+    ///
+    /// static ONCE: SetOnce<u32> = SetOnce::const_new_with(1);
+    ///
+    /// fn get_global_integer() -> Option<&'static u32> {
+    ///     ONCE.get()
+    /// }
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let result = get_global_integer();
+    ///
+    ///     assert_eq!(result, Some(&1));
+    /// }
+    /// ```
+    ///
+    /// [`tokio-console`]: https://github.com/tokio-rs/console
+    /// [unstable feature]: crate#unstable-features
+    #[cfg(not(all(loom, test)))]
+    pub const fn const_new_with(value: T) -> Self {
+        Self {
+            value_set: AtomicBool::new(true),
+            value: UnsafeCell::new(MaybeUninit::new(value)),
+            notify: Notify::const_new(),
+            lock: Mutex::const_new(()),
+        }
+    }
+
+    /// Returns `true` if the `SetOnce` currently contains a value, and `false`
+    /// otherwise.
+    pub fn initialized(&self) -> bool {
+        // Using acquire ordering so we're able to read/catch any writes that
+        // are done with `Ordering::Release`
+        self.value_set.load(Ordering::Acquire)
+    }
+
+    // SAFETY: The SetOnce must not be empty.
+    unsafe fn get_unchecked(&self) -> &T {
+        &*self.value.with(|ptr| (*ptr).as_ptr())
+    }
+
+    /// Returns a reference to the value currently stored in the `SetOnce`, or
+    /// `None` if the `SetOnce` is empty.
+    pub fn get(&self) -> Option<&T> {
+        if self.initialized() {
+            Some(unsafe { self.get_unchecked() })
+        } else {
+            None
+        }
+    }
+
+    // SAFETY: The caller of this function needs to ensure that this function is
+    // called only when the value_set AtomicBool is flipped from FALSE to TRUE
+    // meaning that the value is being set from uinitialized to initialized via
+    // this function
+    //
+    // The caller also has to ensure writes on `value` are syncronized with a
+    // external lock to prevent mutliple set_value calls at the same time.
+    unsafe fn set_value(&self, value: T) {
+        unsafe {
+            self.value.with_mut(|ptr| (*ptr).as_mut_ptr().write(value));
+        }
+
+        // Using release ordering so any threads that read a true from this
+        // atomic is able to read the value we just stored.
+        self.value_set.store(true, Ordering::Release);
+
+        // notify the waiting wakers that the value is set
+        self.notify.notify_waiters();
```

Comment:
Please call `notify_waiters()` after releasing the `self.lock` mutex. This will probably require you to move everything in `set_once` into `set`, but I think that's okay when `set` is so simple already.